File: leetcode/p00203/p00203.go
package p00203

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time Complexity: O(n)
// Space Complexity: O(1)
func removeElements(head *ListNode, val int) *ListNode {
	for head != nil && head.Val == val {
		head = head.Next
	}

	if head == nil {
		return nil
	}

	current := head
	for current.Next != nil {
		if current.Next.Val == val {
			current.Next = current.Next.Next
		} else {
			current = current.Next
		}
	}

	return head
}

File: leetcode/p00203/p00203.py
from typing import Optional, List

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(1)
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        while head and head.val == val:
            head = head.next
        
        if head == None:
            return None
        
        current = head
        while current.next:
            if current.next.val == val:
                current.next = current.next.next
            else:
                current = current.next
        
        return head

def getVals(head: Optional[ListNode]) -> List[int]:
    res = []

    while head:
        res.append(head.val)
        head = head.next

    return res

if __name__ == '__main__':
    s = Solution()
    
    # Example 1
    list = [ListNode(1), ListNode(2), ListNode(6), ListNode(3), ListNode(4), ListNode(5), ListNode(6)]
    for i in range(0, len(list)-1):
        list[i].next = list[i+1]
    assert getVals(s.removeElements(list[0], 6)) == [1,2,3,4,5]

    # Example 2
    assert getVals(s.removeElements(None, 1)) == []

    # Example 3
    list = [ListNode(7), ListNode(7), ListNode(7), ListNode(7)]
    for i in range(0, len(list)-1):
        list[i].next = list[i+1]
    assert getVals(s.removeElements(list[0], 7)) == []